MachOKit/Segments/Sections/MKDataSection.m

```diff
@@ -65,20 +65,17 @@ - (void)dealloc
 //|++++++++++++++++++++++++++++++++++++|//
 - (MKOptional*)childNodeOccupyingVMAddress:(mk_vm_address_t)address targetClass:(Class)targetClass
 {
-    __block MKOptional *child = nil;
-    
-    [_children enumerateKeysAndObjectsUsingBlock:^(__unused NSNumber *key, MKOffsetNode *obj, BOOL *stop) {
-        if ((child = [obj childNodeOccupyingVMAddress:address targetClass:targetClass]).value) {
-            child = [child retain];
-            *stop = YES;
-        } else
-            child = nil;
-    }];
-    
-    if (child)
-        return [child autorelease];
-    else
-        return [super childNodeOccupyingVMAddress:address targetClass:targetClass];
+    // While we could also check whether `address` is _within_ any children, a dict lookup results
+    // in a *net* 60x speed-up, and in most cases `address` directly corresponds to a child anyway.
+    // It might be possible to increase efficiency of the other lookup method using a trie or a
+    // binary search tree but that will probably have a negligible memory improvement and a measurable
+    // impact on performance.
+
```

Comment:
This method should return the correct child if 'address' is *within* a child node.

I suggest that if _children[@(address)] is nil the implementation checks if 'address' is within *this* (self) node and if it is, falls back to the previous linear search.  That will keep the performance gains for the common case while maintaining the existing semantics for the edge cases.

Reply:
I just spent a while looking into this, and I realised there's a bigger problem with `MKDataSection`: the semantics of `-[MKDataSection childNodeOccupyingVMAddress:]` were inherently non-deterministic. As I understand it, the point of `childNodeOccupyingVMAddress` is to find a suitable parent for the child node being created. However, since `MKDataSection` lazily creates child nodes, this results in a problem:

To illustrate this issue, let's say we have an `MKDataSection` which contains nodes of type `MKNodeA` (nodeSize=16), which in turn has a node of type `MKNodeB` (nodeSize=8) at offset 8. The issue is that if we dereference an `MKPointer<MKNodeB>`, it _may or may not_ have `MKNodeA` as its parent depending on whether we've previously dereferenced the `MKNodeA` which "contains" it. This is because `-[MKPointer<MKNodeB> pointee]` would call `-[MKDataSection childNodeAtVMAddress:foo]`, which would create the `MKNodeB` and then call `-[MKDataSection childNodeOccupyingVMAddress:foo]`. Now, if the `MKNodeA` at `[foo - 8, foo + 8)` had been dereferenced previously, the linear search through `_children` would locate it and return it. However, if it had not been created in the past, the linear search would fail and the parent would be set to the `MKDataSection` itself. Thus the expected semantics of `childNodeOccupyingVMAddress` were not being followed in the first place (or at least not always).

In light of this, I realize that my change would make things _more_ deterministic by making it so that an element in an `MKDataSection` always has its parent set to the data section, rather than it depending on whether or not the "real" parent had previously been dereferenced. As far as I can tell, the only way to have determinisim while also maintaining the actual hierarchy would be to non-lazily parse the data section or do some sort of look-behind, however I think it's better to view `MKDataSection` as an unstructured blob because the alternative would require knowledge of the exact format of the data section, and I'm not too sure whether that's well-defined. For example, if an `MKDataSection` is heterogeneous then I don't think it'd even be possible to non-lazily parse it in a way that isn't super hacky.

I did try implementing your suggestion though, and unfortunately the performance is about the same as what it was with a plain linear search, as there are a lot of pointers that point into an `MKDataSection` while parsing classes. It might be possible to overcome this by using an AVL tree to store `_children`, but that would result in a degradation from O(1) to O(log n) performance (in `childNodeOccupyingVMAddress` due to tree searching, and `childNodeAtVMAddress` due to tree insertion), and would additionally retain the non-deterministic behavior of the node's `parent` depending on prior dereferences.

I think the best solution would be to make `MKDataSection`'s children always have the data section as its parent – which I unknowingly already did in my PR, however I can now formalize that. This would result in deterministic behavior plus O(1) lookup time. In fact if this is the defined behavior, another optimization will be possible in that `-[MKDataSection childNodeAtVMAddress:]` can create `targetClass` using `initWithOffset:fromParent:`, skipping the extra step in `initWithVMAddress` to find the parent (as that would, by definition, return the data section anyway).

Reply:
Btw I've committed the aforementioned changes in 583bcf2 in case you'd like to have a look, however I can revert them if necessary.

Reply:
That MKDataSection creates child nodes on-demand is not ideal but is the consequence of there being no way to parse the contents of a data section without information from elsewhere, as you already figured out.

Still thinking about the rest of your comment...

Reply:
`-childNodeOccupyingVMAddress:targetClass:` should maintain the semantics of always returning the *deepest* node in the tree that occupies *address* and is a *targetClass*.  With your current implementation, a client would not be able to say, find the `MKObjCClassMethod` occupying an address because `MKObjCClassMethod` nodes are children of a `MKObjCClassMethodList` node, which is a child of a `MKDataSection` [1].  I understand that this will mean keeping the existing linear search (for now).

`-childNodeAtVMAddress:targetClass:` should follow the same *find the deepest node* semantics of -childNodeOccupyingVMAddress:targetClass:.  My current implementation of `-[MKDataSection childNodeAtVMAddress:targetClass:]` does not.  That and the other flaws you pointed out have led me to conclude overloading `-[MKDataSection childNodeAtVMAddress:targetClass:]` to lazily create child nodes was a mistake.

The current implementation for resolving pointers and lazily creating the pointee nodes should be reworked - possibly by moving to a new method which can be given more suitable and performant semantics.  That can be done later - I will open an issue with more thoughts.  Your changes to `-[MKDataSection childNodeAtVMAddress:targetClass:]` are fine for now.

Reverting the changes to `-[MKDataNode childNodeOccupyingVMAddress:targetClass:]` is the only remaining change needed to this PR.

[1] Or the `MKDataSection` would create the `MKObjCClassMethod` node and add it as a child, which I do not want.  An `MKObjCClassMethod` should be a child of the `MKObjCClassMethodList` that it resides within.

Reply:
Hmm I'm not entirely sure I understand the purpose of reverting the change to `childNodeOccupyingVMAddress`. As far as I can tell, the previous implementation of the aforementioned method had the same flaw as `childNodeAtVMAddress`, in that it was non-deterministic. Even with the change reverted, the semantics would only be correct in the case where the `MKObjCClassMethodList` had already been dereferenced in the past. Otherwise, it would still create the `MKObjCClassMethod` with the `MKDataSection` as its parent.

IMO it would make more sense to revert `childNodeOccupyingVMAddress` once the implementation of `childNodeAtVMAddress` has been changed as you stated, because at that point it would probably also be possible to use a better data structure to remove the need for a linear search (for example, if you decide to non-lazily parse the section somehow, it would make sense to use a sorted array). Reverting this atm would probably not make any improvements to "correctness" (since the method would continue to be non-deterministic), however it would degrade performance significantly. Additionally, I don't think this affects any real-world use cases either, because from what I've seen there are no nodes within `MKDataSection` that have such a hierarchy – for example in the case of `MKObjCClassMethodList`, the element of the list is a pointer (or relative offset on iOS 14) to the method, and not the method itself, so the method would never have the list as its parent anyway.

Reply:
The elements of MKObjCClassMethodList are not pointers.  The runtime defines `method_list_t` as an `entsize_list_tt` of `method_t` structures.

`struct method_list_t : entsize_list_tt<method_t, method_list_t, 0x3>`

Reply:
Oops, I was looking at the definitions of the list sections instead. For the `entsize_list`s, though, aren't the elements explicitly created with parents anyway? https://github.com/DeVaukz/MachO-Kit/blob/20f5fe2645e2707689135e5ee334a678fd947737/MachOKit/ObjC/MKObjCElementList.m#L102

And in case there's a pointer to a method inside a method list from elsewhere, even with my change reverted, whether or not the `MKObjCClassMethod` has the list as its parent will depend on whether the method list has previously been dereferenced.

Reply:
The problem is that with your implementation of `-[MKDataSection childNodeOccupyingVMAddress:targetClass:]`, this code returns the data section instead of the method:

```
let objCClass: MKObjCClass = // Assume this points to a class with at least one method
let objCClassData = objCClass.classData.pointee.value!
let methodList = objCClassData.methods.pointee.value!
let method = methodList.elements.first!
let methodAddress = method.nodeVMAddress
    
let foo = macho.childNodeOccupyingVMAddress(methodAddress, targetClass: nil)
print(foo)
```

With this change:
`__DATA,__objc_const`

With existing implementation:
`<MKObjCClassMethod 0x7ffa6add0b80: address = 0x6a988, size = 0x18>`

Reply:
Ah, I see; would you be okay with cherry-picking to omit that particular change then? I'd like to keep it downstream since I don't need arbitrary access to pointers within `MKDataSection` myself, but I understand why you wouldn't want it upstream for the sake of correctness. 

I'm about to push one more commit which fixes a few memory leaks, but this should be ready to merge once that's done. Also, I haven't had a chance to run all the tests, so you might also want to run those first in case you haven't yet (speaking of which, it's probably worth setting up a CI for that)

Reply:
Or actually nevermind that, I'll just move my `childNodeOccupyingVMAddress` change into another branch so that master is ready to merge without needing to cherry-pick

Reply:
Looks good.  Merging.